lib/history/sqlite.go

```diff
@@ -0,0 +1,224 @@
+/*
+Copyright 2019 Gravitational, Inc.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package history
+
+import (
+	"context"
+	"database/sql"
+	"fmt"
+	"strings"
+	"time"
+
+	pb "github.com/gravitational/satellite/agent/proto/agentpb"
+
+	"github.com/gravitational/trace"
+	_ "github.com/mattn/go-sqlite3" // initialize sqlite3
+)
+
+// SQLiteTimeline represents a timeline of cluster status events. The timeline
+// can hold a specified amount of events and uses a FIFO eviction policy.
+// Timeline events are stored in a local sqlite database.
+//
+// Implements Timeline
+type SQLiteTimeline struct {
+	// size specifies the max number of events stored in the timeline.
+	size int
+	// database points to underlying sqlite database.
+	database *sql.DB
+	// lastStatus holds the last recorded cluster status.
+	lastStatus *Cluster
+}
+
+// NewSQLiteTimeline initializes and returns a new SQLiteTimeline with the
+// specified size. Initial cluster status is `Unknown`.
+func NewSQLiteTimeline(database *sql.DB, size int) (Timeline, error) {
+	if err := initEventTable(database); err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	return &SQLiteTimeline{
+		size:       size,
+		database:   database,
+		lastStatus: &Cluster{Status: pb.SystemStatus_Unknown.String()},
+	}, nil
+}
+
+// initEventTable initializes event table schema.
+func initEventTable(db *sql.DB) error {
+	if _, err := db.Exec(createTableEvent); err != nil {
+		return trace.Wrap(err)
+	}
+	return nil
+}
+
+// RecordStatus records differences of the previous status to the provided
+// status into the Timeline.
+func (t *SQLiteTimeline) RecordStatus(ctx context.Context, status *pb.SystemStatus) error {
+	cluster := parseSystemStatus(status)
+	events := t.lastStatus.diffCluster(cluster)
+	if len(events) == 0 {
+		return nil
+	}
+
+	tx, err := t.database.BeginTx(ctx, nil)
+	if err != nil {
+		return trace.Wrap(err)
+	}
+
+	if err := t.insertEvents(events); err != nil {
+		tx.Rollback()
+		return trace.Wrap(err, "failed to insert events.")
+	}
+
+	if err := t.evictEvents(); err != nil {
```

Comment:
With this approach, the DELETE statement is executed for every inserted event. I wonder if a more efficient approach would be to keep some kind of counter and do a purge every 100 iterations or so. This would make our limit not precise, but reduce number of requests to database. Or do this in a separate goroutine once every few minutes. Or maybe something else.

This is just an optimization though, we can worry about it later.